### PR TITLE
Simplify synapses

### PR DESCRIPTION
**Pull request by oliverrhodes**
Sorry about all the reverted reverts on master - I made a pig's ear of this.

This branch is to simplify the synaptic updates so there are common functions to do the shaping and adding input. This will make it easier to check for saturation of the synaptic input in a single place (coming soon in the precision based weight scaling).

### CODE REVIEW

File: neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h [hunk 1]
```diff
@@ -53,10 +57,24 @@ typedef enum input_buffer_regions {
 //! to the neuron.
 //! \param[in]  parameter: the pointer to the parameters to use
 //! \return nothing
+static inline void _delta_shaping(delta_params_t* delta_params){
```
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h [hunk 2]
```diff
@@ -53,10 +57,24 @@ typedef enum input_buffer_regions {
...
+	_delta_shaping(&parameter->inh);
+
+}
+
+//! \brief helper function to add input for a given timer period to a given
+//! neuron
+//! \param[in]  parameter: the pointer to the parameters to use
+//! \param[in] input the inputs to add.
+//! \return None
+static inline void _add_input_delta(delta_params_t* delta_params, input_t input){
```
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_dual_excitatory_exponential_impl.h [hunk 3]
```diff
@@ -61,18 +62,31 @@ typedef enum input_buffer_regions {
 //! to the neuron.
 //! \param[in]  parameter: the pointer to the parameters to use
 //! \return nothing
+static inline void _exp_shaping(exp_params_t* exp_params){
```
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_dual_excitatory_exponential_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h (hunk 1 above)
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h (hunk 2 above)
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_delta_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_dual_excitatory_exponential_impl.h (hunk 3 above)
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_dual_excitatory_exponential_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_dual_excitatory_exponential_impl.h [hunk 4]
```diff
@@ -61,18 +62,31 @@ typedef enum input_buffer_regions {
...
+	_exp_shaping(&parameter->exc2);
+	_exp_shaping(&parameter->inh);
+}
+
+//! \brief helper function to add input for a given timer period to a given
+//! neuron
+//! \param[in]  parameter: the pointer to the parameters to use
+//! \param[in] input the inputs to add.
+//! \return None
+static inline void _add_input_exp(exp_params_t* exp_params, input_t input){
```
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_dual_excitatory_exponential_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_exponential_impl.h [hunk 5]
```diff
@@ -57,15 +59,30 @@ typedef enum input_buffer_regions {
 //! to the neuron.
 //! \param[in]  parameter: the pointer to the parameters to use
 //! \return nothing
+static inline void _exp_shaping(exp_params_t* exp_params){
```
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_exponential_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

File: neural_modelling/src/neuron/synapse_types/synapse_types_exponential_impl.h [hunk 6]
```diff
@@ -57,15 +59,30 @@ typedef enum input_buffer_regions {
...
+	_exp_shaping(&parameter->exc);
+	_exp_shaping(&parameter->inh);
+}
+
+//! \brief helper function to add input for a given timer period to a given
+//! neuron
+//! \param[in]  parameter: the pointer to the parameters to use
+//! \param[in] input the inputs to add.
+//! \return None
+static inline void _add_input_exp(exp_params_t* exp_params, input_t input){
```
Comment (dkfellows) [inline neural_modelling/src/neuron/synapse_types/synapse_types_exponential_impl.h]: As this isn't Python, prefixing symbol names with `_` isn't recommended.

**Comment by oliverrhodes** [comment]
@dkfellows how's this?

(I have no strong feelings about the underscores, but was simply following what I've seen in other places.)

**Pull request merged**